# Exploration of transit cost data

Through this notebook I hope to Clean the excel file into a usable data set.

1. Remove columns with non-data
2. Impute some values for data with missing years
3. Retain only columns that are relavant for analysis.



In [1]:
# importing the data and checking the data
import pandas as pd
transitcost = pd.read_excel("Data/transitcost.xlsx")
transitcost.head()


ID Country       City        Line        Phase Start year End year  \
0  9501.0      TR   Istanbul          M4   M4 Phase 3       2015     2022   
1  7136.0      CA  Vancouver  Millennium     Broadway       2020     2025   
2  7137.0      CA    Toronto         YUS      Vaughan       2009     2017   
3  7138.0      CA    Toronto         NaN  Scarborough       2020     2030   
4  7139.0      CA    Toronto         NaN      Ontario       2020     2030   

   RR? Length TunnelPer  ... Clength Ctunnel Anglo? Inflation Index  \
0  0.0    7.5         1  ...     7.5     7.5    0.0        1.079102   
1  0.0    5.7  0.877193  ...     0.0     0.0    1.0        1.079102   
2  0.0    8.6         1  ...     0.0     0.0    1.0        1.163176   
3  0.0    7.8         1  ...     0.0     0.0    1.0        1.079102   
4  0.0   15.5      0.57  ...     0.0     0.0    1.0        1.059897   

  Real cost (2021 dollars) Cost/km (2021 dollars) Source2  \
0               550.341886              73.378918    Plan   
1              2565.240650             450.042219   Media   
2              3014.952287             350.575847   Media   
3              4985.450027             639.160260   Media   
4              7632.655004             492.429355    Plan   

                                           Reference  Unnamed: 31  Unnamed: 32  
0  https://www.metro.istanbul/Hatlarimiz/ProjeHal...          NaN          NaN  
1  https://www.translink.ca/Plans-and-Projects/Ra...          NaN          NaN  
2  https://www.thestar.com/news/gta/transportatio...          NaN          NaN  
3  https://urbantoronto.ca/news/2020/03/metrolinx...          NaN          NaN  
4  https://metrolinx.files.wordpress.com/2019/07/...          NaN          NaN  

[5 rows x 33 columns]

In [2]:
transitcost.sample(5)

ID Country                   City                    Line  \
558  9486.0      CN              Guangzhou                 Line 13   
830     NaN      CN               Hangzhou                     NaN   
722     NaN      CN                Qingdao                     NaN   
938     NaN      EG  Alexandria-Port Sa'id  Abo Qir-Port Sa'id LRT   
951     NaN     NaN                    MIN                     NaN   

                      Phase    Start year End year  RR?  Length TunnelPer  \
558         Line 13 Phase 1          2013     2017  0.0   27.03         1   
830             Hangfu Line           NaN     2021  0.0  23.508         1   
722                  Line 9          2021     2026  0.0   16.63         1   
938  Abo Qir-Port Sa'id LRT  Under Tender  By 2026  1.0     280         0   
951                7.793145           NaN      NaN  NaN     NaN       NaN   

     ... Clength Ctunnel Anglo? Inflation Index Real cost (2021 dollars)  \
558  ...     0.0     0.0    0.0        1.163176              6045.460891   
830  ...     0.0     0.0    0.0        1.000000              3178.222000   
722  ...     0.0     0.0    0.0        1.000000              3009.010000   
938  ...     NaN     NaN    NaN             NaN                      NaN   
951  ...     NaN     NaN    NaN             NaN             59940.000000   

    Cost/km (2021 dollars) Source2  \
558             223.657451   Media   
830             135.197465     NaN   
722             180.938665     NaN   
938                    NaN     NaN   
951             149.104478     NaN   

                                             Reference  \
558  http://www.dcement.com/Article/201310/117609.html   
830  http://www.rail-transit.com/project/show.php?i...   
722  http://www.rail-transit.com/project/show.php?i...   
938  http://www.nat.gov.eg/LocationActivity.aspx?id...   
951                                                NaN   

                                           Unnamed: 31  Unnamed: 32  
558                                                NaN          NaN  
830                                                NaN          NaN  
722  https://zfxxgk.ndrc.gov.cn/web/iteminfo.jsp?id...          NaN  
938                                                NaN          NaN  
951                                                NaN          NaN  

[5 rows x 33 columns]

In [3]:
transitcost.tail(5)

ID Country                 City Line           Phase Start year End year  \
960 NaN     NaN   T.COST OF PROJECTS  NaN  4166796.382872        NaN      600   
961 NaN     NaN  AVG COST/KM (W.AVG)  NaN      211.448186        NaN      NaN   
962 NaN     NaN                  NaN  NaN             NaN        NaN      NaN   
963 NaN     NaN         59 Countries  NaN             NaN        NaN      NaN   
964 NaN     NaN           184 Cities  NaN             NaN        NaN      NaN   

     RR?     Length TunnelPer  ... Clength Ctunnel Anglo? Inflation Index  \
960  NaN  12621.074       NaN  ...     NaN     NaN    NaN             NaN   
961  NaN        NaN       NaN  ...     NaN     NaN    NaN             NaN   
962  NaN        NaN       NaN  ...     NaN     NaN    NaN             NaN   
963  NaN        NaN       NaN  ...     NaN     NaN    NaN             NaN   
964  NaN        NaN       NaN  ...     NaN     NaN    NaN             NaN   

    Real cost (2021 dollars) Cost/km (2021 dollars) Source2 Reference  \
960               203.275383                    NaN     NaN       NaN   
961                      NaN                    NaN     NaN       NaN   
962                      NaN                    NaN     NaN       NaN   
963                      NaN                    NaN     NaN       NaN   
964                      NaN                    NaN     NaN       NaN   

     Unnamed: 31  Unnamed: 32  
960          NaN          NaN  
961          NaN          NaN  
962          NaN          NaN  
963          NaN          NaN  
964          NaN          NaN  

[5 rows x 33 columns]

Rows 0 to 943 contain the data. Rest of the rows seem to have been made by the data collectors for analysis. Lets remove these rows. 

In [4]:
transit = transitcost[:944]
transit.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 944 entries, 0 to 943
Data columns (total 33 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   ID                        637 non-null    float64
 1   Country                   944 non-null    object 
 2   City                      944 non-null    object 
 3   Line                      407 non-null    object 
 4   Phase                     943 non-null    object 
 5   Start year                854 non-null    object 
 6   End year                  726 non-null    object 
 7   RR?                       942 non-null    float64
 8   Length                    942 non-null    object 
 9   TunnelPer                 879 non-null    object 
 10  Tunnel                    878 non-null    object 
 11  Elevated                  576 non-null    object 
 12  Atgrade                   570 non-null    object 
 13  Stations                  916 non-null    object 
 14  Platform L

Data type erros - many data type that seem to be contiuous variables seem to be identified as objects. 

In [5]:
# converting column names into lower case
transit.columns = [x.lower() for x in transit.columns.tolist()]
transit.info()

import re
transit.columns = [re.sub(r"\s+", "_", x) for x in transit.columns]
transit.columns = [re.sub(r"/", "_p_", x) for x in transit.columns]
transit.columns = [re.sub(r"[^A-z\s0-9]", "", x) for x in transit.columns]
transit.columns = [re.sub(r"\s+", "_", x) for x in transit.columns]
transit.columns

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 944 entries, 0 to 943
Data columns (total 33 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   id                        637 non-null    float64
 1   country                   944 non-null    object 
 2   city                      944 non-null    object 
 3   line                      407 non-null    object 
 4   phase                     943 non-null    object 
 5   start year                854 non-null    object 
 6   end year                  726 non-null    object 
 7   rr?                       942 non-null    float64
 8   length                    942 non-null    object 
 9   tunnelper                 879 non-null    object 
 10  tunnel                    878 non-null    object 
 11  elevated                  576 non-null    object 
 12  atgrade                   570 non-null    object 
 13  stations                  916 non-null    object 
 14  platform l

Index(['id', 'country', 'city', 'line', 'phase', 'start_year', 'end_year',
       'rr', 'length', 'tunnelper', 'tunnel', 'elevated', 'atgrade',
       'stations', 'platform_length_meters', 'source1', 'cost', 'currency',
       'year', 'ppp_rate', 'real_cost', 'cost_p_km_millions', 'cheap',
       'clength', 'ctunnel', 'anglo', 'inflation_index',
       'real_cost_2021_dollars', 'cost_p_km_2021_dollars', 'source2',
       'reference', 'unnamed_31', 'unnamed_32'],
      dtype='object')

In [6]:
#retaining only useful columns
transit = transit[['country', 'city', 'phase', 'start_year', 'end_year', 'rr', 
                  'length','tunnelper', 'tunnel', 'elevated', 'atgrade', 'stations',
                  'platform_length_meters', 'cost', 'currency', 'year', 'ppp_rate','real_cost', 
                  'cost_p_km_millions', 'anglo', 'inflation_index','real_cost_2021_dollars', 'cost_p_km_2021_dollars']]
transit.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 944 entries, 0 to 943
Data columns (total 23 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   country                 944 non-null    object 
 1   city                    944 non-null    object 
 2   phase                   943 non-null    object 
 3   start_year              854 non-null    object 
 4   end_year                726 non-null    object 
 5   rr                      942 non-null    float64
 6   length                  942 non-null    object 
 7   tunnelper               879 non-null    object 
 8   tunnel                  878 non-null    object 
 9   elevated                576 non-null    object 
 10  atgrade                 570 non-null    object 
 11  stations                916 non-null    object 
 12  platform_length_meters  411 non-null    object 
 13  cost                    906 non-null    object 
 14  currency                910 non-null    ob

## 2. Figuring out missing values

In [7]:
# Print the number of unique values in some columns
print("Unique values:\n")
print(transit.nunique())

# Check for missing data
print("\nMissing values:\n")
print(transit.isnull().sum())


Unique values:

country                    59
city                      193
phase                     599
start_year                 55
end_year                   52
rr                          2
length                    598
tunnelper                 176
tunnel                    486
elevated                   94
atgrade                    63
stations                   56
platform_length_meters     32
cost                      835
currency                   42
year                       43
ppp_rate                  157
real_cost                 872
cost_p_km_millions        882
anglo                       2
inflation_index            37
real_cost_2021_dollars    880
cost_p_km_2021_dollars    882
dtype: int64

Missing values:

country                     0
city                        0
phase                       1
start_year                 90
end_year                  218
rr                          2
length                      2
tunnelper                  65
tunnel                 

### Cost
Most variables are missing values. Let us process the missing variables the following way:
Cost and related variables (real_cost, cost_p_km_millions, real_cost_2021_dollars, cost_p_km_2021_dollars) are all missing same 61 observations. Since cost is a primary variables we are interested in, let us remove all observation with no cost data. 

In [8]:
# subset data for columns with non-null cost variables
transit_cost = transit[transit['real_cost'].notnull()]
transit_cost.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 883 entries, 0 to 943
Data columns (total 23 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   country                 883 non-null    object 
 1   city                    883 non-null    object 
 2   phase                   882 non-null    object 
 3   start_year              806 non-null    object 
 4   end_year                673 non-null    object 
 5   rr                      882 non-null    float64
 6   length                  883 non-null    object 
 7   tunnelper               825 non-null    object 
 8   tunnel                  825 non-null    object 
 9   elevated                524 non-null    object 
 10  atgrade                 522 non-null    object 
 11  stations                860 non-null    object 
 12  platform_length_meters  408 non-null    object 
 13  cost                    883 non-null    object 
 14  currency                883 non-null    ob

## Year

Some years are non-numeric, hence the dataype is object. We want this to be uniform.
There are 
- 213 non-numeric end_year
- 152 end_years after 2023
- 86 non-numeric start_year
- 41 of this 86 non-numeric start years have a numeric end year
- There are 168 numeric start_year values among the non-numeric end years

So, to impute some of these values, I find the average number of years projects have been in existance for each of the country with the missing values, and use the numberic start (end) date and arrive at the end (start) data by subtracting (adding) the average duration of construction. 

In [9]:
# Finding # rows with the issues: 
# non-numeric dates
def non_numeric(col):
    # Identify non-numeric values in the column
    nn = transit_cost[col].apply(lambda x: not str(x).isnumeric())
    print(f"There are {nn.sum()} non-numeric values in {col}")
    transit_cost[nn].head()
non_numeric('start_year')
non_numeric('end_year')

There are 86 non-numeric values in start_year
There are 213 non-numeric values in end_year


In [10]:
# Identify numeric values in the 'End year' column which are greater than 2023
nn_23 = transit_cost['end_year'].apply(lambda x: int(x)>2023 if str(x).isnumeric() else False)
# Count the number of such values
print(f"There are {nn_23.sum()} end_year values > 2023")


There are 152 end_year values > 2023


In [11]:
# identify # rows with end year value but no start year value
nns_nostart = transit_cost['start_year'].apply(lambda x: not str(x).isnumeric()) & transit_cost['end_year'].apply(lambda x: str(x).isnumeric())
print(f"There are {nns_nostart.sum()} numeric end_year values among the non-numeric start years")
# list of countries with this pheonomeon
print("Countries:", transit_cost[nns_nostart].country.unique())
# identify # rows with start year value but no end year value
nns_noend = transit_cost['end_year'].apply(lambda x: not str(x).isnumeric()) & transit_cost['start_year'].apply(lambda x: str(x).isnumeric())
print(f"There are {nns_noend.sum()} numeric start_year values among the non-numeric end years")
# list of countries with this pheonomeon
print("Countries:", transit_cost[nns_noend].country.unique())


There are 41 numeric end_year values among the non-numeric start years
Countries: ['IR' 'CN' 'IT']
There are 168 numeric start_year values among the non-numeric end years
Countries: ['BG' 'DE' 'TH' 'IN' 'TW' 'CN' 'SA' 'BH' 'EG' 'KR' 'CA']


Let us find average built duration for CN, IR, IT and use that to impute start years. 

In [12]:
def avg_dur(country):
    data = transit_cost[transit_cost['country']==country]
    # retaining numeric dates
    numeric = data['start_year'].apply(lambda x: str(x).isnumeric())
    data = data[numeric]
    numeric = data['end_year'].apply(lambda x: str(x).isnumeric())
    data = data[numeric]
    # calculate average duration
    data['start_year'] = pd.to_numeric(data['start_year'], errors='coerce')
    data['end_year'] = pd.to_numeric(data['end_year'], errors='coerce') 
    data['duration'] = data['end_year'] - data['start_year']
    average_duration = data['duration'].mean().round().astype(int)
    # print(f"The average duration of the projects is {average_duration} years for {country}")
    return average_duration
print(f"The average duration of the projects is {avg_dur('CN')} years for CN")
print(f"The average duration of the projects is {avg_dur('IR')} years for IR")
print(f"The average duration of the projects is {avg_dur('IT')} years for IT")

The average duration of the projects is 5 years for CN
The average duration of the projects is 2 years for IR
The average duration of the projects is 10 years for IT


In [13]:
# calcuate average duration for all countries and save into a df - not immediately useful.
country_df = pd.DataFrame(transit_cost['country'].unique())
country_df.rename(columns={0: 'country'}, inplace=True)
country_df['duration'] = country_df['country'].apply (lambda x: avg_dur(x))
country_df.head()

country  duration
0      TR         9
1      CA         7
2      NL        15
3      US         9
4      BG         4

In [14]:
# imputing start years
nns_nostart = transit_cost['start_year'].apply(lambda x: not str(x).isnumeric()) & transit_cost['end_year'].apply(lambda x: str(x).isnumeric())
# condition = transit_cost['start_year'].apply(lambda x: not str(x).isnumeric())
pd.to_numeric(transit_cost['end_year'], errors='coerce')
if nns_nostart.any():
    transit_cost.loc[nns_nostart,'start_year'] = transit_cost['end_year'] - transit_cost.loc[nns_nostart, 'country'].apply(avg_dur)

#imputing end years
nns_noend = transit_cost['end_year'].apply(lambda x: not str(x).isnumeric()) & transit_cost['start_year'].apply(lambda x: str(x).isnumeric())
pd.to_numeric(transit_cost['start_year'], errors='coerce')
if nns_noend.any():
    transit_cost.loc[nns_noend,'end_year'] = transit_cost['start_year'] + transit_cost.loc[nns_noend, 'country'].apply(avg_dur)

In [15]:
# udpated column
transit_cost.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 883 entries, 0 to 943
Data columns (total 23 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   country                 883 non-null    object 
 1   city                    883 non-null    object 
 2   phase                   882 non-null    object 
 3   start_year              845 non-null    object 
 4   end_year                838 non-null    object 
 5   rr                      882 non-null    float64
 6   length                  883 non-null    object 
 7   tunnelper               825 non-null    object 
 8   tunnel                  825 non-null    object 
 9   elevated                524 non-null    object 
 10  atgrade                 522 non-null    object 
 11  stations                860 non-null    object 
 12  platform_length_meters  408 non-null    object 
 13  cost                    883 non-null    object 
 14  currency                883 non-null    ob

From having only 806 non-null start years, and 673 end years, we now have 845 and 838 non-null values respectively.
Next, lets remove any non-numeric and null values from here. 

In [16]:
transit_cost = transit_cost[transit_cost['end_year'].notnull()]
transit_cost['start_year']= transit_cost['start_year'].astype(int)
transit_cost['end_year']=transit_cost['end_year'].astype(int)
transit_cost.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 838 entries, 0 to 943
Data columns (total 23 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   country                 838 non-null    object 
 1   city                    838 non-null    object 
 2   phase                   837 non-null    object 
 3   start_year              838 non-null    int64  
 4   end_year                838 non-null    int64  
 5   rr                      837 non-null    float64
 6   length                  838 non-null    object 
 7   tunnelper               784 non-null    object 
 8   tunnel                  784 non-null    object 
 9   elevated                483 non-null    object 
 10  atgrade                 481 non-null    object 
 11  stations                817 non-null    object 
 12  platform_length_meters  407 non-null    object 
 13  cost                    838 non-null    object 
 14  currency                838 non-null    ob

In [17]:
transit_cost.rename(columns={'real_cost': 'real_cost', 'real_cost_2021_dollars':'real_cost_2021', 'cost_p_km_millions':'cost_km', 'cost_p_km_2021_dollars': 'cost_km_2021' }, inplace=True)
transit_cost.columns

Index(['country', 'city', 'phase', 'start_year', 'end_year', 'rr', 'length',
       'tunnelper', 'tunnel', 'elevated', 'atgrade', 'stations',
       'platform_length_meters', 'cost', 'currency', 'year', 'ppp_rate',
       'real_cost', 'cost_km', 'anglo', 'inflation_index', 'real_cost_2021',
       'cost_km_2021'],
      dtype='object')

In [18]:
transit_df = transit_cost[['country', 'city', 'phase', 'start_year', 'end_year', 
                          'rr', 'length','tunnelper', 'tunnel', 'elevated', 'atgrade', 'stations',
                          'platform_length_meters','anglo','real_cost_2021','cost_km_2021']]

In [19]:
# Sort the 'cost_km_2021' column in descending order
sorted_cost_km_2021 = transit_df['cost_km_2021'].sort_values(ascending=False)
sorted_cost_km_2021[:20].index
transit_df.loc[sorted_cost_km_2021[:20].index]

country           city                      phase  start_year  end_year  \
140      US       New York           East Side Access        2007      2022   
139      US       New York      Second Avenue Phase 2        2019      2029   
138      US       New York      Second Avenue Phase 1        2007      2016   
582      HK      Hong Kong  Tung Chung Line extension        2023      2029   
137      US       New York                7 extension        2007      2014   
141      US       New York                    Gateway        2019      2026   
592      HK      Hong Kong    Modified Initial System        1975      1980   
98       SG      Singapore        Circle Line Stage 6        2017      2025   
583      HK      Hong Kong           West Island Line        2009      2016   
586      HK      Hong Kong    Tseung Kwan O Extension        1999      2002   
142      UK         London                  Crossrail        2009      2021   
73       AU      Melbourne               Metro Tunnel        2018      2025   
16       US  San Francisco           BART to San Jose        2022      2030   
591      HK      Hong Kong       Sha Tin-Central Link        2012      2022   
594      HK      Hong Kong                Island Line        1981      1986   
590      HK      Hong Kong   Tuen Mun South Extension        2023      2030   
152      NZ       Auckland             City Rail Link        2013      2024   
537      EG          Cairo             Line 3 Phase 1        2006      2012   
9        US    Los Angeles             Purple Phase 3        2020      2027   
528      QA           Doha                 Doha Metro        2013      2026   

      rr length tunnelper tunnel elevated atgrade stations  \
140  1.0    2.8         1    2.8        0       0        1   
139  0.0    2.6         1    2.6        0       0        3   
138  0.0    2.7         1    2.7        0       0        3   
582  0.0    1.8         1    1.8        0       0        1   
137  0.0    1.6         1    1.6        0       0        1   
141  1.0    5.3         1    5.3        0       0        0   
592  0.0   15.6     0.821   12.8      2.8       0       15   
98   0.0      4         1      4        0       0        3   
583  0.0      3         1      3        0       0        3   
586  0.0      4         1      4        0       0        5   
142  1.0     21         1     21        0       0        8   
73   0.0      9         1      9        0       0        5   
16   0.0    9.6      0.83      8        0       0        4   
591  1.0     17      0.88     15        0       0        9   
594  0.0   12.5      0.84   10.5        2       0       14   
590  0.0    2.4         0      0        0       0        2   
152  1.0    3.5         1    3.5        0       0        4   
537  0.0    4.3         1    4.3        0       0        5   
9    0.0    4.2         1    4.2        0       0        2   
528  0.0    124  0.899194  111.5        0       0      NaN   

    platform_length_meters  anglo  real_cost_2021  cost_km_2021  
140                    NaN    1.0    12575.764608   4491.344503  
139                    187    1.0     6949.000000   2672.692308  
138                    187    1.0     5542.538066   2052.791876  
582                    NaN    0.0     3559.561327   1977.534071  
137                    NaN    1.0     2982.389845   1863.993653  
141                    NaN    1.0     9500.000000   1792.452830  
592                    200    0.0    23080.745707   1479.534981  
98                     NaN    1.0     5480.500000   1370.125000  
583                    NaN    0.0     4004.347296   1334.782432  
586                    200    0.0     4748.426667   1187.106667  
142                    NaN    1.0    21332.155179   1015.816913  
73                     NaN    1.0     8683.236000    964.804000  
16                     NaN    1.0     9100.000000    947.916667  
591                    NaN    0.0    16058.686841    944.628638  
594                    200    0.0    11678.490324  

## Imputing values for: Phase, Regional Rail

There is one project with no phase name, and one with no info on if it is a rr or not. Lets give the former the same name as the city, and latter 

In [20]:
# for phase, lets give the phase name as the city name
print(transit_df[transit_df['phase'].isnull()])
transit_df.at[943,'phase']=str(transit_df.at[943,'city'])+str(' Line')
print(transit_df.loc[943])

    country         city phase  start_year  end_year   rr length tunnelper  \
943      RO  Cluj-Napoca   NaN        2023      2031  0.0     21         1   

    tunnel elevated atgrade stations platform_length_meters  anglo  \
943     21      NaN     NaN       19                    NaN    1.0   

     real_cost_2021  cost_km_2021  
943          4446.0    211.714286  
country                                 RO
city                           Cluj-Napoca
phase                     Cluj-Napoca Line
start_year                            2023
end_year                              2031
rr                                     0.0
length                                  21
tunnelper                                1
tunnel                                  21
elevated                               NaN
atgrade                                NaN
stations                                19
platform_length_meters                 NaN
anglo                                  1.0
real_cost_2021              

In [21]:
# for regional rail, none of the metros in Soudi are regional rail
print(transit_df[transit_df['rr'].isnull()])
transit_df.at[526,'rr']=0.0
print(transit_df.loc[526])

    country   city        phase  start_year  end_year  rr length tunnelper  \
526      SA  Mecca  Mecca Metro        2009      2014 NaN    188       NaN   

    tunnel elevated atgrade stations platform_length_meters  anglo  \
526    NaN      NaN     NaN       90                    NaN    0.0   

     real_cost_2021  cost_km_2021  
526     47459.76373    252.445552  
country                            SA
city                            Mecca
phase                     Mecca Metro
start_year                       2009
end_year                         2014
rr                                0.0
length                            188
tunnelper                         NaN
tunnel                            NaN
elevated                          NaN
atgrade                           NaN
stations                           90
platform_length_meters            NaN
anglo                             0.0
real_cost_2021            47459.76373
cost_km_2021               252.445552
Name: 526, dtype: obje

## Tunnel details and number of station

1. convert places with no tunnels - assign city wise avegate tunnel perc and multiply by the lenght of line for which tunnel is missing
2. for station# missing = find countrywise avg station per km of line, ans assign to missing row *length of the row


In [22]:
print(transit_df.isnull().sum())

country                     0
city                        0
phase                       0
start_year                  0
end_year                    0
rr                          0
length                      0
tunnelper                  54
tunnel                     54
elevated                  355
atgrade                   357
stations                   21
platform_length_meters    431
anglo                       0
real_cost_2021              0
cost_km_2021                0
dtype: int64


In [23]:
def non_numeric(col):
    # Identify non-numeric values in the column
    nn = transit_df[col].isnull()
    print(f"There are {nn.sum()} non-numeric values in {col}")
    return nn
df = non_numeric('tunnelper')

# print(f'There are {transit_df[df].city.nunique()} are in the following cities:', transit_df[df].city.unique())
tunnelgroup = transit_df.groupby('city')['tunnelper'].mean()

# this cleans 
for index, row in transit_df[df].iterrows():
    transit_df.loc[index, 'tunnel'] =(tunnelgroup[transit_df.loc[index, 'city']])* transit_df.loc[index,'length'] 
    transit_df.loc[index, 'tunnelper'] = transit_df.loc[index, 'tunnel']/transit_df.loc[index, 'length']

non_numeric('tunnelper')

transit_df = transit_df[transit_df['tunnelper'].notnull()]

non_numeric('tunnelper')

There are 54 non-numeric values in tunnelper
There are 3 non-numeric values in tunnelper
There are 0 non-numeric values in tunnelper


0      False
1      False
2      False
3      False
4      False
       ...  
905    False
906    False
907    False
908    False
943    False
Name: tunnelper, Length: 835, dtype: bool

In [24]:
print(transit_df.isnull().sum())

country                     0
city                        0
phase                       0
start_year                  0
end_year                    0
rr                          0
length                      0
tunnelper                   0
tunnel                      0
elevated                  352
atgrade                   354
stations                   21
platform_length_meters    429
anglo                       0
real_cost_2021              0
cost_km_2021                0
dtype: int64


In [25]:
def no_station(col):
    # Identify non-numeric values in the column
    nn = transit_df[col].isnull()
    print(f"There are {nn.sum()} non-numeric values in {col}")
    return nn
    # print(transit_df[nn])
stations = no_station('stations')
stationdf = transit_df.groupby('country')[['length','stations']].mean()
# there is one country with missing station data for all its projects - Qatar. We find the number as available online replace the value there
stationdf.at['QA','stations']=37
# finding number of stations per km
stationdf['st_km']= (stationdf['stations']/stationdf['length'])

There are 21 non-numeric values in stations


In [26]:
for index, row in transit_df[stations].iterrows():
    transit_df.at[index,'stations'] =  row['length'] * stationdf.loc[row['country'], 'st_km']
print(transit_df.isnull().sum())

country                     0
city                        0
phase                       0
start_year                  0
end_year                    0
rr                          0
length                      0
tunnelper                   0
tunnel                      0
elevated                  352
atgrade                   354
stations                    0
platform_length_meters    429
anglo                       0
real_cost_2021              0
cost_km_2021                0
dtype: int64


In [28]:
transit_df.to_csv("Data/transit_cleaned.csv")